File: frontend/src/types/entities/Order.ts
import { type Address } from "./Client";

export interface OrderItem {
  id: number;
  name: string;
  quantity: number;
  price_per_item: number;
  product_id: string | null;
}

export interface Order {
  order_id: string;
  client_id: string;
  client_name: string;
  date: string;
  time_preference: string;
  time_slot?: string;
  delivery_phone: string;
  delivery_address: Address;
  items: OrderItem[];
  payment_method?: string;
  comment?: string;
  note?: string;
  is_paid: boolean;
  recurring_order_id?: string | null;
}

File: backend/src/backend/domain/services/order.py
from datetime import date, time
from decimal import Decimal

from backend.application.vars import (
    ClientId,
    OrderId,
    PaymentMethod,
    ProductId,
    ShopId,
)
from backend.infrastructure.persistence.tables.orders import (
    Order,
    OrderItem,
)


def create_order(
    *,
    order_id: OrderId,
    shop_id: ShopId,
    client_id: ClientId,
    delivery_date: date,
    delivery_start_time: time,
    delivery_end_time: time,
    delivery_phone: str,
    delivery_address: dict,
    payment_method: PaymentMethod,
    comment: str | None,
) -> Order:
    return Order(
        id=order_id,
        date=delivery_date,
        delivery_address=delivery_address,
        delivery_phone=delivery_phone,
        delivery_start_time=delivery_start_time,
        delivery_end_time=delivery_end_time,
        comment=comment,
        shop_id=shop_id,
        client_id=client_id,
        payment_method=payment_method,
    )


def create_order_item(
    *,
    name: str,
    quantity: int,
    price_per_item: Decimal,
    product_id: ProductId | None = None,
) -> OrderItem:
    return OrderItem(
        name=name,
        quantity=quantity,
        price_per_item=price_per_item,
        product_id=product_id,
    )
